File: standalone/src/main/scala/io/delta/standalone/internal/util/PartitionUtils.scala
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.delta.standalone.internal.util

import java.util.Locale

import scala.collection.JavaConverters._

import io.delta.standalone.expressions.{And, Expression, Literal}
import io.delta.standalone.types.StructType

import io.delta.standalone.internal.actions.AddFile
import io.delta.standalone.internal.data.PartitionRowRecord


private[internal] object PartitionUtils {

  /**
   * Filters the given [[AddFile]]s by the given `partitionFilters`, returning those that match.
   *
   * This is different from
   * [[io.delta.standalone.internal.scan.FilteredDeltaScanImpl.getFilesScala]] in that this method
   * already has the [[AddFile]]s in memory, whereas the `FilteredDeltaScanImpl` performs a
   * memory-optimized replay to collect and filter the files.
   *
   * @param files The active files in the DeltaLog state, which contains the partition value
   *              information
   * @param partitionFilter Filter on the partition columns
   */
  def filterFileList(
      partitionSchema: StructType,
      files: Seq[AddFile],
      partitionFilter: Expression): Seq[AddFile] = {
    files.filter { addFile =>
      val partitionRowRecord = new PartitionRowRecord(partitionSchema, addFile.partitionValues)
      val result = partitionFilter.eval(partitionRowRecord)
      result.asInstanceOf[Boolean]
    }
  }

  /**
   * Partition the given condition into two optional conjunctive predicates M, D such that
   * condition = M AND D, where we define:
   * - M: conjunction of predicates that can be evaluated using metadata only.
   * - D: conjunction of other predicates.
   */
  def splitMetadataAndDataPredicates(
      condition: Expression,
      partitionColumns: Seq[String]): (Option[Expression], Option[Expression]) = {
    val (metadataPredicates, dataPredicates) = splitConjunctivePredicates(condition)
      .partition(isPredicateMetadataOnly(_, partitionColumns))

    val metadataConjunction = if (metadataPredicates.isEmpty) {
      None
    } else {
      Some(metadataPredicates.reduceLeftOption(new And(_, _)).getOrElse(Literal.True))
    }

    val dataConjunction = if (dataPredicates.isEmpty) {
      None
    } else {
      Some(dataPredicates.reduceLeftOption(new And(_, _)).getOrElse(Literal.True))
    }

    (metadataConjunction, dataConjunction)
  }

  /**
   * Check if condition can be evaluated using only metadata (i.e. partition columns)
   */
  def isPredicateMetadataOnly(condition: Expression, partitionColumns: Seq[String]): Boolean = {
    val lowercasePartCols = partitionColumns.map(_.toLowerCase(Locale.ROOT))

    condition.references()
      .asScala
      .map(_.toLowerCase(Locale.ROOT))
      .forall(lowercasePartCols.contains(_))
  }

  private def splitConjunctivePredicates(condition: Expression): Seq[Expression] = {
    condition match {
      case a: And => splitConjunctivePredicates(a.getLeft) ++ splitConjunctivePredicates(a.getRight)
      case other => other :: Nil
    }
  }

}

File: kernel/kernel-api/src/main/java/io/delta/kernel/utils/PartitionUtils.java
/*
 * Copyright (2024) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.delta.kernel.utils;

import static java.util.Objects.requireNonNull;

import io.delta.kernel.Scan;
import io.delta.kernel.Snapshot;
import io.delta.kernel.data.FilteredColumnarBatch;
import io.delta.kernel.data.Row;
import io.delta.kernel.engine.Engine;
import io.delta.kernel.expressions.Predicate;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.HashSet;
import java.util.Set;

public class PartitionUtils {

  private PartitionUtils() {}

  /**
   * Check if a partition exists (i.e. actually has data) in the given {@link Snapshot} based on the
   * given {@link Predicate}.
   *
   * @param engine the {@link Engine} to use for scanning the partition.
   * @param snapshot the {@link Snapshot} to scan.
   * @param partitionPredicate the {@link Predicate} to use for filtering the partition.
   * @return true if the partition exists, false otherwise.
   * @throws IllegalArgumentException if the predicate does not reference any partition columns or
   *     if it references any data columns
   */
  public static boolean partitionExists(
      Engine engine, Snapshot snapshot, Predicate partitionPredicate) {
    requireNonNull(engine, "engine is null");
    requireNonNull(snapshot, "snapshot is null");
    requireNonNull(partitionPredicate, "partitionPredicate is null");

    final Set<String> snapshotPartColNames = new HashSet<>(snapshot.getPartitionColumnNames());

    io.delta.kernel.internal.util.PartitionUtils.validatePredicateOnlyOnPartitionColumns(
        partitionPredicate, snapshotPartColNames);

    final Scan scan = snapshot.getScanBuilder().withFilter(partitionPredicate).build();

    try (CloseableIterator<FilteredColumnarBatch> columnarBatchIter = scan.getScanFiles(engine)) {
      while (columnarBatchIter.hasNext()) {
        try (CloseableIterator<Row> selectedRowsIter = columnarBatchIter.next().getRows()) {
          if (selectedRowsIter.hasNext()) {
            return true;
          }
        }
      }
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }

    return false;
  }
}
